### PR TITLE
[Concept Entry] PyTorch Tensors

### LINKED ISSUE #4825: [Concept Entry] PyTorch Tensors
### Reason/inspiration (optional)

We would like a new entry on the `Tensors` concept under PyTorch. The entry should go in a new file under `docs/content/pytorch/concepts/tensors/tensors.md`.

The entry should include:

- An introduction to the concept
- A ‘Syntax’ section that provides the syntax for the concept (if applicable)
- An ‘Example’ section that provides an example demonstrating the concept in use
- A ‘Codebyte Example’ section that provides a codebyte example demonstrating the concept in use

Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by noahpgordon**
### Description

Created **tensors.md** within pytorch/concepts.

### Issue Solved

Closes #4825 

### Type of Change

- Adding a new entry

### Checklist

- [X] All writings are my own.
- [X] My entry follows the Codecademy Docs style guide.
- [X] My changes generate no new warnings.
- [X] I have performed a self-review of my own writing and code.
- [X] I have checked my entry and corrected any misspellings.
- [X] I have made corresponding changes to the documentation if needed.
- [X] I have confirmed my changes are not being pushed from my forked `main` branch.
- [X] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [X] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/pytorch/concepts/tensors/tensors.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+print(myTensor.shape)
+```
+
+Output:
+
+```shell
+torch.Size([8, 2, 8, 1])
+torch.Size([128])
+torch.Size([64, 2])
+```
```
Comment (SaviDahegaonkar) [inline content/pytorch/concepts/tensors/tensors.md:219]:
``` suggestion
Output:

``` shell
torch.Size([8, 2, 8, 1])
torch.Size([128])
torch.Size([64, 2])
```
```
The output should be preceded by a, sentence like this `This code gives the following output.` or something similar to this sentence.

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @noahpgordon,
I have suggested you some changes regarding your file. Please get it updated as soon as possible and let me know.
Thanks.

File: content/pytorch/concepts/tensors/tensors.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  - 'AI'
+Tags:
+  - 'AI'
+  - 'Data Structures'
+  - 'Arrays'
+  - 'Deep Learning'
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'py-torch-for-classification'
+  - 'paths/machine-learning'
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:16]: I added the missing paths to catalog-content.md.

File: content/pytorch/concepts/tensors/tensors.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+---
+
+**Tensors** are multi-dimensional arrays containing values of the same data type. These types include booleans, floats, and integers. Tensors are a fundamental data structure in PyTorch. When building a machine learning model in PyTorch, all input data, output data, and model weights are stored as tensors.
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:19]: Your suggested edit doesn't seem clearer to me.

File: content/pytorch/concepts/tensors/tensors.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+### Data-Specific Tensors
+
+If the values within a tensor do matter, use the following methods:
+
+- For all zeros: `torch.zeros()`
+- For all ones: `torch.ones()`
+- For specified values: `torch.tensor()`
+- For all randomly-generated values: `torch.rand()`
+
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:61]: Should we really put all examples in every section within their own 'example' subsection? It seems to me that this would be unproductive, as examples are integrated into the exposition in several sections. I've written concept entries for this project before, and don't recall this being a standard. Also, here are some examples within the Python section that don't use this convention:

https://www.codecademy.com/resources/docs/python
https://www.codecademy.com/resources/docs/python/conditionals
https://www.codecademy.com/resources/docs/python/strings
https://www.codecademy.com/resources/docs/python/generators
https://www.codecademy.com/resources/docs/python/decorators

File: content/pytorch/concepts/tensors/tensors.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  - 128-bit: `dtype=torch.complex128`
+- Signed and unsigned integers
+  - 8-bit: `dtype=torch.int8` and `dtype=torch.uint8`
+  - 16-bit: `dtype=torch.int16` and `dtype=torch.uint16`
+  - 32-bit: `dtype=torch.int32` and `dtype=torch.uint32`
+  - 64-bit: `dtype=torch.int64` and `dtype=torch.uint64`
+- Quantized integers
+  - 8-bit: `dtype=torch.qint8`
+  - 32-bit: `dtype=torch.qint32`
+- Booleans: `dtype=torch.bool`
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:112]: The official PyTorch documentation states that both torch.complex32 and torch.complex64 are supported datatypes:

https://pytorch.org/docs/stable/tensors.html

File: content/pytorch/concepts/tensors/tensors.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+#Create a float16 tensor by explicit declaration
+myExFloat16Tensor = torch.tensor([[3.4, 0.9], [2.3, 1.3]], dtype=torch.float16)
+
+#Convert the float16 tensor to an int16 tensor using .to()
+myExFloat16Tensor = myExFloat16Tensor.to(torch.int16)
+
+print(myBoolTensor,"\n")
+print(myExFloat16Tensor)
+```
+
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:133]: Do you really think it's more readable to say "this code prints the following output", vs "output"? It seems unnecessarily wordy to me.

File: content/pytorch/concepts/tensors/tensors.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+print(myTensor.shape)
+```
+
+Output:
+
+```shell
+torch.Size([2, 4, 6, 7])
+```
+
+Operations with multiple tensors, like adding or multiplying tensors, generally require the tensors involved to have the same shapes.
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:165]: I use notes in-line with the style guidance to qualify previous statements that the reader may be mislead by. This sentence is not a qualification, it's a statement that tells the reader why tensor shapes are important.

File: content/pytorch/concepts/tensors/tensors.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+```shell
+tensor([[ True, False],
+        [False, False]]) 
+
+tensor([[3, 0],
+        [2, 1]], dtype=torch.int16)
+```
+
+Notice that when printing a tensor with a numerical datatype other than the default `torch.float32`, the datatype will be specified in the output.
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md:144]: This is also not a note. It's not a qualification of something that may be misleading, it's simply drawing attention to a feature exhibited by the example. (Just because the word "notice" is similar to "note" does not mean they are synonymous.)

**Comment by noahpgordon** [comment]
@SaviDahegaonkar I have incorporated some of your suggestions and left my thoughts on the others.

**Comment by SaviDahegaonkar** [comment]
> @SaviDahegaonkar I have incorporated some of your suggestions and left my thoughts on the others.

Will check on it and let you know asap.

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @noahpgordon ,
I am sorry for the misunderstanding by my side. LGTM! I hereby approve this.


Thanks,
Savi

File: content/pytorch/concepts/tensors/tensors.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+print(myBoolTensor,"\n")
+print(myExFloat16Tensor)
+```
+
+This results in the following output:
+
+```shell
+tensor([[ True, False],
+        [False, False]])
```
[the remark this answers is not in the document]
Comment (noahpgordon) [reply content/pytorch/concepts/tensors/tensors.md]: I've committed this suggestion.

**Comment by avdhoottt** [comment]
Hey @noahpgordon, sorry, but we have to close this PR as one is already present for tensor operations. I apologize for the mistake. You can get assigned for some other issue. Thank you!

**Comment by noahpgordon** [comment]
> Hey @noahpgordon, sorry, but we have to close this PR as one is already present for tensor operations. I apologize for the mistake. You can get assigned for some other issue. Thank you!

I don't understand. This entry is a general overview of Tensors in PyTorch and is broader than the entry on Tensor Operations, which is presumably why these were created as two distinct issues. It covers many issues not covered on the operations entry, including creating and initializing tensors, tensor datatypes, and tensor shapes. It does not cover tensor operations in detail, in fact it actually links to the Tensor Operations page.

If the team wants only one page, why not merge the two, for example putting the content currently on Tensor Operations as a section of this page?

If there is going to be no page on Tensors generally but only Tensor Operations, where will the term entries like [.size()](https://github.com/Codecademy/docs/pull/5088) and [.view()](https://github.com/Codecademy/docs/pull/5215) go? It wouldn't make sense to put them all as term entries under Tensor Operations, since not all of them are operations on tensors. I'll also note that these term entries currently use a path file that uses this concept entry.

**Comment by avdhoottt** [comment]
@noahpgordon, I understand what you are saying. I forgot to mention that I'll merge those two entries and give you credit as a co-author. We'll move the files under the Tensor operations concept entry. Thanks!!

**Comment by noahpgordon** [comment]
@avdhoottt That's fine, but wouldn't it make more sense to rename the Tensor Operations entry to Tensors then? There will be material on the concept entry and term entries under it that are not about tensor operations.

**Comment by noahpgordon** [comment]
@avdhoottt There are currently 8 term entries in the codebase that are invisible on the main site because this PR was closed, with the newest being added 3 days ago and the oldest 2 months ago. That's 8 people writing entries and 16 reviews, not including this PR. The original justification for closing this still makes no sense, since this page and the Tensor Operations page are clearly very distinct.

Looping in the other reviewers for this PR as well @SaviDahegaonkar @jrood

**Comment by jrood** [comment]
@noahpgordon My sincere apologies for the misunderstanding that happened here. We definitely do need this concept entry, parent to the term entries mentioned. I'll coordinate with the team to make sure we get this PR re-opened and merged.

**Review by avdhoottt — APPROVED** [review]
LGTM!!
